clr: SWDEV-547890 - Maintain an MQD for the emulated AQL queue

lancesix opened the pull request:
## Motivation

This PR re-submits changes introduced by #1316, which have had to be reverted because of conflicts with #1183.  This new PR is the same as the original one, plus the fix to address the conflicts with "SWDEV-556684 - Remove HSAIL support (#1183)".

## Technical Details

To simplify the shader debugger implementation, maintain the relevant parts of the emulated AQL queue's MQD (amd_queue_t): read_dispatch_id, write_dispatch_id, compute_tmpring_size.

With this MQD, the shader debugger can handle the emulated AQL queue the same way it does the real AQL queue, no specialization is required.

Conservatively update the MQD's read_dispatch_id. The read_dispatch_id cannot be smaller than the current aql_packet_id - hsa_queue.size for the debugger to work correctly.

The read_dispatch_id really should be updated when the CmdBuf is marked as complete. Left a FIXME to address it in a future commit.

## Test Plan

Test debugger functionality.

## Test Result

Debugger works as expected, simulated AQL queues are handles as native ones.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

lmoriche committed:
clr: SWDEV-547890 - Conservatively update the MQD's read_dispatch_id

The read_dispatch_id cannot be smaller than the current aql_packet_id
- hsa_queue.size for the debugger to work correctly.

The read_dispatch_id really should be updated when the CmdBuf is marked
as complete. Left a FIXME to address it in a future commit.

ammallya commented:
/AzurePipelines run rocm-ci-caller

gandryey commented on projects/clr/rocclr/device/pal/palvirtual.cpp:
```diff
@@ -2729,9 +2762,46 @@ bool VirtualGPU::submitKernelInternal(const amd::NDRangeContainer& sizes, const
+  // Do an atomic max of &amd_queue.read_dispatch_id and new_read_dispatch_id
+  uint64_t old_read_dispatch_id = amd_queue.read_dispatch_id;
+  while (new_read_dispatch_id > old_read_dispatch_id) {
+#if defined(__GNUC__)
```
Let's switch to c++ atomics

lancesix replied:
I can switch to c++20's atomic_ref, which looks way better.

```c++
// It would be ill-formed to have an atomic_ref of a volatile
// type if is_always_lock_free is false.
static_assert (std::atomic_ref<decltype (mgmt.amd_queue_.write_dispatch_id)>::is_always_lock_free);
uint64_t packet_id = std::atomic_ref (mgmt.amd_queue_.write_dispatch_id).fetch_add(1, std::memory_order::relaxed);
```

However in this case, MVSC gives unexpected warnings with `read_dispatch_id` being volatile (see https://godbolt.org/z/E3asjTPP4 for an example).

This should be valid, and more recent MSVC seem to be happy with such construct (starting msvc v19.41 - see https://godbolt.org/z/q9WeEqeWh), but in the meantime I am reluctant adding code that would generate all of those warnings.

lancesix replied:
Done, the new patch now uses std::atomic_ref instead

lancesix committed:
clr: SWDEV-547890 - Maintain an MQD for the emulated AQL queue

To simplify the shader debugger implementation, maintain the relevant
parts of the emulated AQL queue's MQD (amd_queue_t): read_dispatch_id,
write_dispatch_id, compute_tmpring_size.

With this MQD, the shader debugger can handle the emulated AQL queue
the same way it does the real AQL queue, no specialization is required.
